### PR TITLE
Ta activity stats

### PR DESCRIPTION
**Pull request by finnergizer**
Current stats provided for admins and graders in the submission status column providing info on assignments completed and average number of annotations per submission that has been marked as having been 'completed' marking.

See https://slack-files.com/T0AGA2MQX-F0CPFN7QE-cabb89d8c8 for details.

Also see https://slack-files.com/T0AGA2MQX-F0DMA2NRX-5d7dc9c897 for more details on TA grade distribution

### CODE REVIEW

File: app/models/admin.rb [hunk 1]
```diff
@@ -7,4 +7,41 @@ class Admin < User
...
 
+  # for an admin, we just want all the assignment groupings submitted
+  def get_num_assigned(assignment)
+    assignment.groupings.size
+  end
+
+  # for an admin, we want all the assignment groupings marked completed
+  def get_num_marked(assignment)
+    n = 0
+    assignment.groupings.each do |x|
```
Comment (david-yz-liu) [inline app/models/admin.rb]: I think you can use the `count` method here

File: app/models/ta.rb [hunk 2]
```diff
@@ -29,6 +29,28 @@ def get_num_marked(assignment)
...
+  end
+
+  def average_annotations(assignment)
+    num_marked = get_num_marked(assignment)
+    avg = 0
+    if num_marked != 0
+      num_annotations = get_num_annotations(assignment)
+      avg = num_annotations / num_marked
+    end
+    avg
```
Comment (david-yz-liu) [inline app/models/ta.rb]: You should round this.

File: config/locales/en.yml [hunk 3]
```diff
@@ -455,6 +455,7 @@ en:
       tags_list:            "Filter by Tag:"
       search_submissions:   "Search Submissions"
       how_many_marked:      "%{num_marked}/%{num_assigned} assignments have been marked"
+      average_annotations:  "%{average_annotations} annotations per marked submission"
```
Comment (david-yz-liu) [inline config/locales/en.yml]: Please add this to the other locale files.

Comment (finnergizer) [inline config/locales/en.yml]: Added and translated. Thanks!

File: app/views/assignments/_list_manage.html.erb [hunk 4]
```diff
@@ -7,77 +7,105 @@
...
+                                action: :browse,
+                                assignment_id: assignment.id %>
+
+                </span></li>
+                <% end %>
+                <% if !assignment.results_average.nil? %>
+                  <li><span id='class_average__<%= assignment.short_identifier %>'>
+
                     <%= t(:class_average,
                           results_average: assignment.results_average) %>
```
Comment (arkon) [inline app/views/assignments/_list_manage.html.erb]: I think it'd be nice to have a % sign afterwards (assuming it's a percentage average). @david-yz-liu : thoughts?

Comment (david-yz-liu) [inline app/views/assignments/_list_manage.html.erb]: I agree. Also, @finnergizer can you make sure this is rounded to two decimal places?

Comment (finnergizer) [inline app/views/assignments/_list_manage.html.erb]: Added, and also checked that rounding is to two decimal places when it is calculated server-side.

File: app/controllers/main_controller.rb [hunk 5]
```diff
@@ -160,10 +160,21 @@ def index
     ]).order('id ASC')
 
     @current_assignment = Assignment.get_current_assignment
+    @current_ta = @current_assignment.tas.first
 
     render :index, layout: 'content'
   end
 
+  def grader_summary
```
Comment (david-yz-liu) [inline app/controllers/main_controller.rb]: This should be a method in the `Assignment` controller

File: app/controllers/tas_controller.rb [hunk 6]
```diff
@@ -95,4 +95,17 @@ def upload_ta_list
   def user_params
     params.require(:user).permit(:user_name, :last_name, :first_name)
   end
+
+  before_action :set_assignment_ta, only: [:refresh_graph]
```
Comment (david-yz-liu) [inline app/controllers/tas_controller.rb]: Don't make a separate method; put the two lines inside the `refresh_graph` method.

File: app/controllers/tas_controller.rb [hunk 7]
```diff
@@ -95,4 +95,17 @@ def upload_ta_list
...
+
+  def set_assignment_ta
+    @assignment = Assignment.find(params[:assignment])
+    @current_ta = Ta.find(params[:id])
+  end
+
+  def refresh_graph
+    respond_to do |format|
```
Comment (david-yz-liu) [inline app/controllers/tas_controller.rb]: We generally haven't used `respond_to` with only one format; you can remove this.

File: app/models/admin.rb [hunk 8]
```diff
@@ -7,4 +7,36 @@ class Admin < User
   after_destroy :revoke_repository_permissions
   after_update  :maintain_repository_permissions
 
+  # for an admin, we just want all the assignment groupings submitted
+  def get_num_assigned(assignment)
```
Comment (david-yz-liu) [inline app/models/admin.rb]: Make this a method in the `Assignment` model.

File: app/models/admin.rb [hunk 9]
```diff
@@ -7,4 +7,36 @@ class Admin < User
...
+    assignment.groupings.size
+  end
+
+  # for an admin, we want all the assignment groupings marked completed
+  def get_num_marked(assignment)
+    n = assignment.groupings.count(marking_completed: true)
```
Comment (david-yz-liu) [inline app/models/admin.rb]: Don't use a temporary variable here.

File: app/models/admin.rb [hunk 10]
```diff
@@ -7,4 +7,36 @@ class Admin < User
...
+    n
+  end
+
+  def get_num_annotations(assignment)
+    n = 0
+    assignment.groupings.each do |x|
+      # only grab annotations from groupings where marking is completed
+      next unless x.marking_completed?
+      x.submissions.each do |s|
```
Comment (david-yz-liu) [inline app/models/admin.rb]: This one is interesting - use `current_submission_used` here.

File: app/models/admin.rb [hunk 11]
```diff
@@ -7,4 +7,36 @@ class Admin < User
...
+
+  # for an admin, we want all the assignment groupings marked completed
+  def get_num_marked(assignment)
+    n = assignment.groupings.count(marking_completed: true)
+    n
+  end
+
+  def get_num_annotations(assignment)
+    n = 0
+    assignment.groupings.each do |x|
```
Comment (david-yz-liu) [inline app/models/admin.rb]: You should be able to use the [sum](http://apidock.com/rails/Enumerable/sum) method here.

File: app/models/ta.rb [hunk 12]
```diff
@@ -6,9 +6,9 @@ class Ta < User
   CSV_UPLOAD_ORDER = USER_TA_CSV_UPLOAD_ORDER
   SESSION_TIMEOUT = USER_TA_SESSION_TIMEOUT
 
-  after_create  :grant_repository_permissions
```
Comment (david-yz-liu) [inline app/models/ta.rb]: Revert these changes.

File: app/models/ta.rb [hunk 13]
```diff
@@ -29,6 +29,28 @@ def get_num_marked(assignment)
     n
   end
 
+  def get_num_annotations(assignment)
```
Comment (david-yz-liu) [inline app/models/ta.rb]: Okay, so the three methods above - make an optional parameter by a ta_id, which filters for a specific TA.

File: app/models/ta.rb [hunk 14]
```diff
@@ -50,24 +72,48 @@ def get_criterion_associations_by_assignment(assignment)
 
   def get_criterion_associations_count_by_assignment(assignment)
     assignment.criterion_ta_associations
-              .where(ta_id: id)
```
Comment (david-yz-liu) [inline app/models/ta.rb]: Revert all of these changes. You shouldn't be changing style at all.

File: app/models/ta.rb [hunk 15]
```diff
@@ -50,24 +72,48 @@ def get_criterion_associations_by_assignment(assignment)
...
+      distribution = update_distribution(distribution, result,
+                                         assignment.total_mark, intervals)
+    end # end of groupings loop
+    distribution.to_json
+  end
+
+  def update_distribution(distribution, result, out_of, intervals)
+    steps = 100 / intervals # number of percentage steps in each interval
+    percentage = (result.total_mark / out_of * 100).ceil
+    if percentage == 0 then distribution[0] += 1
```
Comment (david-yz-liu) [inline app/models/ta.rb]: Put bodies on different lines than the conditions.

File: app/views/assignments/_assignment_summary.html.erb [hunk 16]
```diff
@@ -14,13 +14,19 @@
...
         <%= link_to I18n.t(:refresh_graph),
                     refresh_graph_assignment_path(id: assignment.id),
                     remote: true %>
        </p>
     </div>
+    <br>
+    <div id='assignment_<%= assignment.id %>_ta_graph' class='left'>
+      <%= render partial: 'graders/ta_grade_distribution_graph', locals: { assignment: assignment, ta: ta } %>
+    </div>
+    <div class="right">
```
Comment (david-yz-liu) [inline app/views/assignments/_assignment_summary.html.erb]: Use single-quotes

File: app/views/assignments/_grader_summary.html.erb [hunk 17]
```diff
@@ -1,16 +1,39 @@
-<p class="grader_distribution"><%= I18n.t(:grader_distribution ) %></p>
+<style>
```
Comment (david-yz-liu) [inline app/views/assignments/_grader_summary.html.erb]: This should be in the CSS, not here.

File: app/views/assignments/_grader_summary.html.erb [hunk 18]
```diff
@@ -1,16 +1,39 @@
...
+
+  ::-webkit-scrollbar-thumb {
+    border-radius: 4px;
+    background-color: rgba(0, 0, 0, .5);
+    -webkit-box-shadow: 0 0 1px rgba(255, 255, 255, .5);
+  }
+</style>
 
 <% if assignment.tas.empty? %>
+  <p class="grader_distribution"><%= I18n.t(:grader_distribution) %></p>
```
Comment (david-yz-liu) [inline app/views/assignments/_grader_summary.html.erb]: Inside the views, omit `I18n`. Also, use single-quotes.

File: app/views/assignments/_grader_summary.html.erb [hunk 19]
```diff
@@ -1,16 +1,39 @@
...
-               num_marked: ta.get_num_marked(assignment),
-               num_assigned: ta.get_num_assigned(assignment)) %>)
-      </li>
-    <% end %>
-  </ul>
+  <p class="grader_distribution"><%= I18n.t(:grader_distribution) %>
+    (<%= link_to t('graders.all_graders'), { :controller => "main", :action => "grader_summary",
+                                             :assignment => assignment.id }
+  %>)</p>
+  <div style="height:380px; overflow:scroll">
```
Comment (david-yz-liu) [inline app/views/assignments/_grader_summary.html.erb]: Put this in the CSS.

File: app/views/assignments/_grader_summary.html.erb [hunk 20]
```diff
@@ -1,16 +1,39 @@
...
-    <% assignment.tas.each do |ta| %>
-      <li><%= "#{ta.first_name} #{ta.last_name}" %>
-        (<%= t('browse_submissions.how_many_marked',
-               num_marked: ta.get_num_marked(assignment),
-               num_assigned: ta.get_num_assigned(assignment)) %>)
-      </li>
-    <% end %>
-  </ul>
+  <p class="grader_distribution"><%= I18n.t(:grader_distribution) %>
+    (<%= link_to t('graders.all_graders'), { :controller => "main", :action => "grader_summary",
```
Comment (david-yz-liu) [inline app/views/assignments/_grader_summary.html.erb]: Use the new style of hashes, e.g. `controller: 'main'` Also use single-quotes.

File: app/views/assignments/_list_manage.html.erb [hunk 21]
```diff
@@ -7,77 +7,105 @@
     <div class='table'>
       <table>
         <thead>
-          <tr>
-            <th><%= t(:assignment_item) %></th>
-            <th><%= t(:due) %></th>
-            <th><%= t(:submission_status) %></th>
-          </tr>
+        <tr>
```
Comment (david-yz-liu) [inline app/views/assignments/_list_manage.html.erb]: I'm not going to review this part until you change the indentation back to how it was before.

File: app/views/assignments/view_summary.js.erb [hunk 22]
```diff
@@ -1,11 +1,15 @@
...
-                                locals: { assignment: @assignment }).html_safe %>";
+                                locals: { assignment: @assignment, ta: @current_ta}).html_safe %>";
 
 jQuery('#assignment_<%= @assignment.id %>_graph').html(
   "<%= escape_javascript(render partial: 'grade_distribution_graph',
                                 locals: { assignment: @assignment }) %>");
 
+jQuery('#assignment_<%= @assignment.id %>_ta_graph').html(
+    "<%= escape_javascript(render partial: 'graders/ta_grade_distribution_graph',
+                                locals: { assignment: @assignment, ta: @current_ta }) %>");
```
Comment (david-yz-liu) [inline app/views/assignments/view_summary.js.erb]: The indentation is off here.

File: app/views/graders/_ta_grade_distribution_graph.html.erb [hunk 23]
```diff
@@ -0,0 +1,35 @@
+<% unless ta.nil? %>
+  <h3><%= ta.first_name + " " + ta.last_name%></h3>
+  <div>
+    (<%= t('browse_submissions.how_many_marked',
+           num_marked: ta.get_num_marked(@assignment),
```
Comment (david-yz-liu) [inline app/views/graders/_ta_grade_distribution_graph.html.erb]: You seem to be passing in a local variable `assignment` here, which is good - you should not use the `@assignment` in this case.

File: app/views/graders/_ta_grade_distribution_graph.html.erb [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+        strokeColor: 'rgba(151,187,205,1)',
+        data:        <%= ta.grade_distribution_array(assignment,20) %>
+      }]
+    };
+
+    var options = {
+      scaleLineWidth: 2,
+      barStrokeWidth: 1,
+      barValueSpacing: 2,
+      scaleLabel: "   <%%=value%>"
```
Comment (david-yz-liu) [inline app/views/graders/_ta_grade_distribution_graph.html.erb]: I'm not really sure what's going on here. Is only one `%` necessary?

File: app/views/main/grader_summary.html.erb [hunk 25]
```diff
@@ -0,0 +1,26 @@
+<% content_for :head do %>
```
Comment (david-yz-liu) [inline app/views/main/grader_summary.html.erb]: This file should be under `assignments` not `main`

File: app/views/main/grader_summary.html.erb [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+  <%= javascript_include_tag 'Chart' %>
+<% end %>
+
+<div class='title_bar'>
+  <h1>
+    <%= t(:assignment_link,
+            short_id: @assignment.short_identifier,
```
Comment (david-yz-liu) [inline app/views/main/grader_summary.html.erb]: The indentation here is off.

File: app/views/main/grader_summary.html.erb [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+            description: @assignment.description) %> -
+    <%=t(:grader_distribution)%>
```
Comment (david-yz-liu) [inline app/views/main/grader_summary.html.erb]: There should be spaces here.

File: app/views/main/grader_summary.html.erb [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+  </h1>
+</div>
+<div class='wrapper'>
+  <div id='section'>
+    <div style="text-align: center">
+      <% @assignment.tas.each do |ta| %>
+        <div style="display:inline-block;">
+          <div id='assignment_<%= @assignment.id %>_ta_
```
Comment (david-yz-liu) [inline app/views/main/grader_summary.html.erb]: This `id` shouldn't be split onto multiple lines

File: app/views/main/grader_summary.html.erb [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+            <% ta.id %>_graph'>
+            <%= render partial: 'graders/ta_grade_distribution_graph', locals: { assignment: @assignment, ta: ta, a:1 } %>
```
Comment (david-yz-liu) [inline app/views/main/grader_summary.html.erb]: On the other hand, this line is too long. I'm also not sure what `a` means.

File: app/views/graders/_ta_grade_distribution_graph.html.erb [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+        strokeColor: 'rgba(151,187,205,1)',
+        data:        <%= ta.grade_distribution_array(assignment,20) %>
+      }]
+    };
+
+    var options = {
+      scaleLineWidth: 2,
+      barStrokeWidth: 1,
+      barValueSpacing: 2,
+      scaleLabel: '   <%%=value%>'
```
Comment (finnergizer) [inline app/views/graders/_ta_grade_distribution_graph.html.erb]: This is an option that is passed that prevents the labels on the Y-axis from appearing slightly cut off. I needed the additional % sign for it to work.

File: app/models/ta.rb (hunk 15 above)
Comment (finnergizer) [inline app/models/ta.rb]: I was getting hound errors for the method being > 10 lines so I compacted the one-line bodies. That's also why I split up the `grade_distribution_array` method into two methods as well.

File: app/assets/stylesheets/main.scss [hunk 31]
```diff
@@ -82,6 +82,23 @@
...
     margin-bottom: 0;
   }
+
+  #grader-right {
+    height: 380px;
+    overflow: scroll;
+  }
+
+  ::-webkit-scrollbar {
+    -webkit-appearance: none;
```
Comment (david-yz-liu) [inline app/assets/stylesheets/main.scss]: Yeah I don't know about this one. If @arkon knows what the correct solution is great, if not I'm fine ignoring it.

File: app/models/ta.rb [hunk 32]
```diff
@@ -15,13 +15,13 @@ class Ta < User
...
   has_many :grade_entry_students, through: :grade_entry_student_tas
 
-  def get_num_assigned(assignment)
-    assignment.ta_memberships.where(user_id: id).size
+  def get_num_assigned(assignment, ta_id = id)
+    assignment.ta_memberships.where(user_id: ta_id).size
   end
 
-  def get_num_marked(assignment)
+  def get_num_marked(assignment, ta_id = id)
```
Comment (david-yz-liu) [inline app/models/ta.rb]: Make these methods only appear in the the `Assignment` model. Give it a signature `get_num_marked(assignment, ta_id = nil)` so that admins can call it without specifying an id and get everything, and tas can call it with their id to get the numbers specific to them.

**Comment by david-yz-liu** [comment]
@finnergizer I'm quite happy with this pull request now. I've made a couple of comments, but once those are done I'm confident about merging this in.

File: app/assets/stylesheets/main.scss (hunk 31 above)
Comment (arkon) [inline app/assets/stylesheets/main.scss]: This is fine since it is indeed WebKit-specific. The box-shadow a few lines down should be prefix-free, however.

**Comment by finnergizer** [comment]
Thanks for the help @david-yz-liu. Made the suggested logic change and added a check for an edge case. I also refactored the assignment stat methods to be in the assignment model only, as well as removed the one vendor prefix suggested.

File: app/models/assignment.rb [hunk 33]
```diff
@@ -756,6 +756,55 @@ def groups_submitted
     groupings.select(&:has_submission?)
   end
 
+  def get_num_assigned(assignment, ta_id = nil)
```
Comment (david-yz-liu) [inline app/models/assignment.rb]: This is better, but one thing I forgot to mention last time is that these are all instance methods, so passing in an extra `assignment` argument is pretty redundant.

Basically just remove all references to `assignment` in these methods. It will still work the same way.

Comment (finnergizer) [inline app/models/assignment.rb]: Yea I noticed this as well actually when I was changing it, but wasn't sure if I should stick to the signatures. I've made the changes now :).

**Comment by david-yz-liu** [comment]
Great work, @finnergizer!
